installinator/src/block_size_writer.rs

```diff
@@ -0,0 +1,223 @@
+    /// calling (1) `flush()` or (2) `shutdown()` will write any
+    /// buffered-but-not-yet-written data to the underlying buffer regardless of
+    /// its length.
+    pub(crate) struct BlockSizeWriter<W> {
+        #[pin]
+        inner: W,
+        buf: Vec<u8>,
+        block_size: usize,
+    }
+}
+
+impl<W: AsyncWrite> BlockSizeWriter<W> {
+    pub(crate) fn with_block_size(block_size: usize, inner: W) -> Self {
+        Self { inner, buf: Vec::with_capacity(block_size), block_size }
+    }
+
+    #[cfg(test)]
+    fn into_inner(self) -> W {
+        self.inner
+    }
+
+    fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut W> {
+        self.project().inner
+    }
+
+    fn flush_buf(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<io::Result<()>> {
+        let mut me = self.project();
+        let mut written = 0;
+        let mut ret = Ok(());
+        while written < me.buf.len() {
+            match ready!(me.inner.as_mut().poll_write(cx, &me.buf[written..])) {
+                Ok(0) => {
+                    ret = Err(io::Error::new(
+                        io::ErrorKind::WriteZero,
+                        "failed to write the buffered data",
+                    ));
+                    break;
+                }
+                Ok(n) => written += n,
+                Err(e) => {
+                    ret = Err(e);
+                    break;
+                }
+            }
+        }
+        if written > 0 {
+            me.buf.drain(..written);
+        }
+        Poll::Ready(ret)
+    }
+}
+
+impl<W: AsyncWrite> AsyncWrite for BlockSizeWriter<W> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        // We should never buffer up more than `block_size` data.
+        assert!(self.buf.len() <= self.block_size);
+
+        // If we already have exactly `block_size` bytes, begin by flushing.
+        if self.buf.len() == self.block_size {
+            ready!(self.as_mut().flush_buf(cx))?;
+        }
+
+        let me = self.project();
+        if me.buf.is_empty() {
+            // If our buffer is empty, either directly write `block_size` bytes
+            // from `buf` to `inner` (if there's enough data in `buf`) or just
+            // copy it into our buffer.
+            if buf.len() >= *me.block_size {
+                me.inner.poll_write(cx, &buf[..*me.block_size])
+            } else {
+                // `me.buf` is empty and `buf` is strictly less than
+                // `self.block_size`, so just copy it.
+                me.buf.extend_from_slice(buf);
+                Poll::Ready(Ok(buf.len()))
+            }
+        } else {
+            // Our buffer already has data - just copy as much of `buf` as we
+            // can onto the end of it.
+            let n = usize::min(*me.block_size - me.buf.len(), buf.len());
+            me.buf.extend_from_slice(&buf[..n]);
```

Comment:
After looking at this again, I see that we may fill up a block size of data, but if so, we don't necessarily write it to the underlying writer via a call to `me.inner.poll_write`. If this is the last chunk of data though, are we guaranteed to get another call to `self.poll_write` or a call to `poll_flush`?

Reply:
We're not; the onus is on the caller to call `.flush()` when they're done. tokio's `BufWriter` and `std::io::BufWriter` have the same requirements; tokio's docs include

> When the BufWriter is dropped, the contents of its buffer will be discarded... If you need to write out the contents of its buffer, you must manually call flush before the writer is dropped.

I could add something similar to the block comment on `BlockSizeWriter`?

Reply:
This makes sense. Thank you! I think a block comment would be warranted.

Reply:
:+1: Added in [ad4f011](https://github.com/oxidecomputer/omicron/pull/2950/commits/ad4f01147dcd0966799fac039346b8c3ffd3420f)